File: code/java/ch11/src/main/java/com/github/zxh0/luago/api/LuaState.java
package com.github.zxh0.luago.api;

public interface LuaState {

    int LUA_MINSTACK = 20;
    int LUAI_MAXSTACK = 1000000;
    int LUA_REGISTRYINDEX = -LUAI_MAXSTACK - 1000;
    long LUA_RIDX_GLOBALS = 2;

    /* basic stack manipulation */
    int getTop();
    int absIndex(int idx);
    boolean checkStack(int n);
    void pop(int n);
    void copy(int fromIdx, int toIdx);
    void pushValue(int idx);
    void replace(int idx);
    void insert(int idx);
    void remove(int idx);
    void rotate(int idx, int n);
    void setTop(int idx);
    /* access functions (stack -> Go); */
    String typeName(LuaType tp);
    LuaType type(int idx);
    boolean isNone(int idx);
    boolean isNil(int idx);
    boolean isNoneOrNil(int idx);
    boolean isBoolean(int idx);
    boolean isInteger(int idx);
    boolean isNumber(int idx);
    boolean isString(int idx);
    boolean isTable(int idx);
    boolean isThread(int idx);
    boolean isFunction(int idx);
    boolean isJavaFunction(int idx);
    boolean toBoolean(int idx);
    long toInteger(int idx);
    Long toIntegerX(int idx);
    double toNumber(int idx);
    Double toNumberX(int idx);
    String toString(int idx);
    JavaFunction toJavaFunction(int idx);
    /* push functions (Go -> stack); */
    void pushNil();
    void pushBoolean(boolean b);
    void pushInteger(long n);
    void pushNumber(double n);
    void pushString(String s);
    void pushJavaFunction(JavaFunction f);
    void pushJavaClosure(JavaFunction f, int n);
    void pushGlobalTable();
    /* comparison and arithmetic functions */
    void arith(ArithOp op);
    boolean compare(int idx1, int idx2, CmpOp op);
    /* get functions (Lua -> stack) */
    void newTable();
    void createTable(int nArr, int nRec);
    LuaType getTable(int idx);
    LuaType getField(int idx, String k);
    LuaType getI(int idx, long i);
    LuaType getGlobal(String name);
    /* set functions (stack -> Lua) */
    void setTable(int idx);
    void setField(int idx, String k);
    void setI(int idx, long i);
    void setGlobal(String name);
    void register(String name, JavaFunction f);
    /* 'load' and 'call' functions (load and run Lua code) */
    ThreadStatus load(byte[] chunk, String chunkName, String mode);
    void call(int nArgs, int nResults);
    /* miscellaneous functions */
    void len(int idx);
    void concat(int n);

}

File: code/csharp/ch07/luavm/api/LuaState.cs
using System;
using LuaType = System.Int32;
using ArithOp = System.Int32;
using CompareOp = System.Int32;

namespace luavm.api
{
    public partial interface LuaState
    {
        int LuaType { get; set; }
        int GetTop();
        int AbsIndex(int idx);
        bool CheckStack(int n);
        void Pop(int n);
        void Copy(int fromIdx, int toIdx);
        void PushValue(int idx);
        void Replace(int idx);
        void Insert(int idx);
        void Remove(int idx);
        void Rotate(int idx, int n);
        void SetTop(int idx);

        /* access functions (stack->c#) */
        string TypeName(LuaType tp);
        LuaType Type(int idx);
        bool IsNone(int idx);
        bool IsNil(int idx);
        bool IsNoneOrNil(int idx);
        bool IsBoolean(int idx);
        bool IsInteger(int idx);
        bool IsNumber(int idx);
        bool IsString(int idx);
        bool ToBoolean(int idx);
        long ToInteger(int idx);
        Tuple<long, bool> ToIntegerX(int idx);
        double ToNumber(int idx);
        Tuple<double, bool> ToNumberX(int idx);
        string ToString(int idx);
        Tuple<string, bool> ToStringX(int idx);

        /* push functions (c# -> stack) */
        void PushNil();
        void PushBoolean(bool b);
        void PushInteger(long n);
        void PushNumber(double n);
        void PushString(string s);

        void Arith(ArithOp op);
        bool Compare(int idx1, int idx2, CompareOp op);
        void Len(int idx);
        void Concat(int n);

        void NewTable();
        void CreateTable(int nArr, int nRec);
        LuaType GetTable(int idx);
        LuaType GetField(int idx, string k);
        LuaType GetI(int idx, long i);
        void SetTable(int idx);
        void SetField(int idx, string k);
        void SetI(int idx, long n);
    }
}